src/commands/azureServiceOperators/helpers/servicePrincipalHelper.ts

```diff
@@ -0,0 +1,148 @@
+    if (failed(spInfo)) {
+        return spInfo;
+    }
+
+    const cloudName = spInfo.result.session.environment.name;
+    const tenantId = spInfo.result.session.tenantId;
+    const promiseResults = await Promise.all(apiAzureAccount.filters.map(f => getSubscriptionAccess(spInfo.result.credential, f.subscription, spInfo.result)));
+    const ownershipResults = combine(promiseResults);
+    if (failed(ownershipResults)) {
+        return ownershipResults;
+    }
+
+    const subscriptions = ownershipResults.result.filter(r => r.hasRoleAssignment).map(r => ({
+        id: r.subscription.subscriptionId || "",
+        name: r.subscription.displayName || ""
+    }));
+
+    return { succeeded: true, result: {cloudName, tenantId, subscriptions} };
+}
+
+async function getServicePrincipalInfo(sessions: AzureSession[], appId: string, appSecret: string): Promise<Errorable<ServicePrincipalInfo>> {
+    const spInfoResults = await Promise.all(sessions.map(s => getServicePrincipalInfoForSession(s, appId, appSecret)));
+    for (const spInfoResult of spInfoResults) {
+        if (succeeded(spInfoResult)) {
+            return spInfoResult;
+        }
+    }
+
+    const spInfosResult = combine(spInfoResults);
+    if (succeeded(spInfosResult)) {
+        // Can only happen if there were no sessions, otherwise we would've returned success above.
+        return { succeeded: false, error: 'No Azure sessions found.' };
+    }
+
+    return spInfosResult;
+}
+
+async function getServicePrincipalInfoForSession(session: AzureSession, appId: string, appSecret: string): Promise<Errorable<ServicePrincipalInfo>> {
+    const baseUrl = getGraphClientBaseUrl(session.environment);
+    const graphClientOptions: TokenCredentialAuthenticationProviderOptions = {
+        scopes: [`${baseUrl}/.default`]
+    };
+
+    const credential = new ClientSecretCredential(session.tenantId, appId, appSecret);
+
+    const graphClient = GraphClient.initWithMiddleware({
+        baseUrl,
+        authProvider: new TokenCredentialAuthenticationProvider(credential, graphClientOptions)
+    });
+
+    let spSearchResults: any;
+    try {
+        spSearchResults = await graphClient.api("/servicePrincipals").filter(`appId eq '${appId}'`).select(['id', 'displayName']).get();
+    } catch (e) {
+        return { succeeded: false, error: `Failed to retrieve service principal: ${getErrorMessage(e)}` };
+    }
+
+    if (!spSearchResults.value || spSearchResults.value.length !== 1) {
+        return { succeeded: false, error: `Expected service principal search result to contain value with one item. Actual result: ${JSON.stringify(spSearchResults)}` };
+    }
+
+    const searchResult = spSearchResults.value[0];
+    const spInfo = {
+        id: searchResult.id,
+        displayName: searchResult.displayName,
+        session,
+        credential
+    };
+
+    return { succeeded: true, result: spInfo };
+}
+
+function getGraphClientBaseUrl(environment: Environment): string {
```

Comment:
💡 Thinking outlaud: I wonder if we rename this to something more contextual? although the api is graph (Just sharing some thought homz)

Reply:
As an aside, I've always found the name 'graph' for the API kind of confusing...

Is `getMicrosoftGraphClientBaseUrl` any better? It's a bit long, but it distinguishes it from Azure AD Graph at least. Any thoughts/suggestions?